File: code/typescript/src/nodes.ts
/**
 * Clase base abstracta para nodos genéricos.
 * @template T Tipo del valor almacenado en el nodo
 */
export abstract class Node<T> {
  /**
   * Valor almacenado en el nodo
   */
  public value: T;

  /**
   * Crea una instancia de Node
   * @param value Valor inicial del nodo
   */
  constructor(value: T) {
    this.value = value;
  }

  toString(): string {
    return `(${this.value})`;
  }
}

/**
 * Clase base abstracta para vértices de grafos con funcionalidad de adyacencia
 * @template T Tipo del valor almacenado en el vértice
 * @template Adjacency Tipo de las adyacencias para definir un vertice con peso
 * y uno sin peso
 */
export abstract class Vertex<T, Adjacency> extends Node<T> {
  public lvl: number | undefined;
  /**
   * Indica si el vértice ha sido visitado en un recorrido
   */
  public visited: boolean;
  /**
   * Lista de vértices adyacentes
   */
  public adjacencies: Adjacency[];

  /**
   * Crea una instancia de Vertex
   * @param value Valor del vértice
   * @param lvl Nivel
   * @param adjacencies Lista inicial de adyacencias
   */
  public constructor({
    value,
    lvl = undefined,
    adjacencies = undefined,
  }: {
    value: T;
    lvl?: number;
    adjacencies?: Adjacency[];
  }) {
    super(value);
    this.lvl = lvl;
    this.adjacencies = adjacencies ? adjacencies : [];
    this.visited = false;
  }

  /**
   * Añade una o más adyacencias al vértice
   * @param adjacencies Adyacencias a añadir
   */
  public append(...adjacencies: Adjacency[]) {
    adjacencies.forEach((adjacency) => {
      this.adjacencies.push(adjacency);
    });
  }

  /**
   * Obtiene una adyacencia por índice
   * @param index Posición en la lista de adyacencias
   * @returns Adyacencia en la posición solicitada o undefined
   */
  public get(index: number): Adjacency | undefined {
    return this.adjacencies[index];
  }
}

/**
 * Nodo para listas doblemente enlazadas
 * @template T Tipo del valor almacenado en el nodo
 */
export class DoubleLinkedNode<T> extends Node<T> {
  /**
   * Referencia al siguiente nodo en la lista
   */
  public next: DoubleLinkedNode<T> | undefined;
  /**
   * Referencia al nodo anterior en la lista
   */
  public prev: DoubleLinkedNode<T> | undefined;

  /**
   * Crea una instancia de DoubleLinkedNode
   * @param value Valor del nodo
   * @param next Referencia al siguiente nodo (opcional)
   * @param prev Referencia al nodo anterior (opcional)
   */
  public constructor(
    value: T,
    next: DoubleLinkedNode<T> | undefined = undefined,
    prev: DoubleLinkedNode<T> | undefined = undefined,
  ) {
    super(value);
    this.next = next;
    this.prev = prev;
  }
}

/**
 * Vértice para grafos no ponderados
 * @template T Tipo del valor almacenado en el vértice
 */
export class NonWeightedVertex<T> extends Vertex<T, NonWeightedVertex<T>> {
  /**
   * Crea una instancia de NonWeightedVertex
   * @param value Valor del vértice
   * @param lvl Nivel
   * @param adjacencies Lista inicial de vértices adyacentes
   */
  public constructor({
    value,
    lvl = undefined,
    adjacencies = undefined,
  }: {
    value: T;
    lvl?: number;
    adjacencies?: NonWeightedVertex<T>[];
  }) {
    super({ value, lvl, adjacencies });
  }
}

/**
 * Vértice para grafos ponderados
 * @template T Tipo del valor almacenado en el vértice
 */
export class WeightedVertex<T> extends Vertex<T, [WeightedVertex<T>, number]> {
  /**
   * Crea una instancia de WeightedVertex
   * @param value Valor del vértice
   * @param lvl Nivel
   * @param adjacencies Lista inicial de adyacencias ponderadas
   */
  public constructor({
    value,
    lvl = undefined,
    adjacencies = undefined,
  }: {
    value: T;
    lvl?: number;
    adjacencies?: [WeightedVertex<T>, number][];
  }) {
    super({ value, lvl, adjacencies });
  }
}

File: code/rust/src/nodes.rs
//! Módulo para implementar una lista doblemente enlazada usando Rc y RefCell

use std::cell::RefCell;
use std::fmt;
use std::rc::{Rc, Weak};

/// Nodo de una lista doblemente enlazada
///
/// # Campos
/// - `value`: Valor almacenado en el nodo (genérico)
/// - `next`: Referencia al siguiente nodo (fuertemente tipada con Rc)
/// - `prev`: Referencia al nodo anterior (débilmente tipada con Weak)
#[derive(Debug, Default, Clone)]
pub struct Node<T> {
    /// Valor almacenado en el nodo
    pub value: T,

    /// Referencia al siguiente nodo en la lista
    ///
    /// Usa `Rc<RefCell<...>>` para permitir:
    /// - Múltiples propietarios
    /// - Mutabilidad interior
    pub next: Option<Rc<RefCell<Node<T>>>>,

    /// Referencia al nodo anterior en la lista
    ///
    /// Usa `Weak<RefCell<...>>` para:
    /// - Evitar ciclos de referencia
    /// - Permitir la liberación de memoria
    pub prev: Option<Weak<RefCell<Node<T>>>>,
}

impl<T: fmt::Display> fmt::Display for Node<T> {
    /// Implementa la representación en cadena del nodo
    ///
    /// # Ejemplo
    /// ```
    /// let node = Node { value: 42, next: None, prev: None };
    /// assert_eq!(format!("{}", node), "(42)");
    /// ```
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({})", self.value)
    }
}

/// Estado de valor no definido para el patrón Builder
#[derive(Debug, Clone, Default)]
pub struct NoValue;

/// Contenedor para el valor definido del nodo
#[derive(Debug, Clone, Default)]
pub struct Value<T>(T);

/// Builder para crear nodos de forma segura y escalonada
///
/// # Tipos genéricos
/// - `T`: Tipo del valor almacenado en el nodo
/// - `V`: Tipo de estado del valor (NoValue | Value<T>)
///
/// # Patrón de diseño
/// Usa el patrón Builder con tipos fantasma para garantizar:
/// - Asignación obligatoria del valor
/// - Configuración flexible de referencias
pub struct NodeBuilder<T, V> {
    /// Valor del nodo (en estado NoValue o Value<T>)
    pub value: V,

    /// Referencia al siguiente nodo
    pub next: Option<Rc<RefCell<Node<T>>>>,

    /// Referencia al nodo anterior
    pub prev: Option<Weak<RefCell<Node<T>>>>,
}

impl<T> NodeBuilder<T, NoValue> {
    /// Crea un nuevo NodeBuilder en estado inicial (sin valor)
    ///
    /// # Ejemplo
    /// ```
    /// let builder = NodeBuilder::<i32, _>::new();
    /// ```
    pub fn new() -> NodeBuilder<T, NoValue> {
        NodeBuilder {
            value: NoValue::default(),
            next: None,
            prev: None,
        }
    }
}

impl<T> NodeBuilder<T, Value<T>> {
    /// Construye el nodo final consumiendo el Builder
    ///
    /// # Retorno
    /// `Rc<RefCell<Node<T>>>` - Nodo listo para usar en la lista
    ///
    /// # Ejemplo
    /// ```
    /// let node = NodeBuilder::new()
    ///     .value(42)
    ///     .build();
    /// ```
    pub fn build(self) -> Rc<RefCell<Node<T>>> {
        let node = Node {
            value: self.value.0,
            next: self.next,
            prev: self.prev,
        };

        Rc::new(RefCell::new(node))
    }
}

impl<T, V> NodeBuilder<T, V> {
    /// Establece el valor del nodo
    ///
    /// # Argumentos
    /// - `value`: Valor a almacenar en el nodo
    ///
    /// # Ejemplo
    /// ```
    /// let builder = NodeBuilder::new().value(42);
    /// ```
    pub fn value(self, value: T) -> NodeBuilder<T, Value<T>> {
        NodeBuilder {
            value: Value(value),
            next: self.next,
            prev: self.prev,
        }
    }

    /// Establece la referencia al siguiente nodo
    ///
    /// # Argumentos
    /// - `next`: Referencia al siguiente nodo (Rc<RefCell<Node<T>>>)
    ///
    /// # Nota
    /// Esta implementación no actualiza automáticamente la referencia
    /// `prev` del nodo siguiente.
    ///
    /// # Ejemplo
    /// ```
    /// let node1 = NodeBuilder::new().value(1).build();
    /// let node2 = NodeBuilder::new().value(2).next(node1).build();
    /// ```
    pub fn next(self, next: Rc<RefCell<Node<T>>>) -> NodeBuilder<T, V> {
        NodeBuilder {
            value: self.value,
            next: Some(next),
            prev: self.prev,
        }
    }

    /// Establece la referencia al nodo anterior
    ///
    /// # Argumentos
    /// - `prev`: Referencia débil al nodo anterior (Weak<RefCell<Node<T>>>)
    ///
    /// # Ejemplo
    /// ```
    /// let node1 = NodeBuilder::new().value(1).build();
    /// let node2 = NodeBuilder::new()
    ///     .value(2)
    ///     .prev(Rc::downgrade(&node1))
    ///     .build();
    /// ```
    pub fn prev(self, prev: Weak<RefCell<Node<T>>>) -> NodeBuilder<T, V> {
        NodeBuilder {
            value: self.value,
            next: self.next,
            prev: Some(prev),
        }
    }
}

/// Ejemplo básico de uso
///
/// # Nota
/// Esta implementación no mantiene automáticamente la consistencia
/// bidireccional de los enlaces entre nodos.
pub fn example_usage() {
    // Crear nodo base
    let node = NodeBuilder::new().value(1).build();

    // Crear segundo nodo enlazado al primero
    let node2 = NodeBuilder::new().value(2).next(node).build();

    // Crear tercer nodo con referencia al segundo
    let _node3 = NodeBuilder::new().value(3).prev(Rc::downgrade(&node2));
}
